feat(sqllab): non-blocking persistence mode

justinpark opened the pull request:
### SUMMARY
Following up [SIP-93](https://github.com/apache/superset/issues/21385), this commit implements the first step of non-blocking persistence mode for the current working sqllab editor activities.

As described in [SIP-93](https://github.com/apache/superset/issues/21385), the existing PERSISTENCE logic causes a user interaction delay due to the network (or server side) lagging. For example, user had to wait the server response in order to update the tab name or query limit count. (see the before screenshot)

Therefore, this commit aims to improve current tab activities by replacing the following blocking logics with a non-blocking way by cumulating the changes and syncing them with the latest updates asynchronously: 

```
- querySuccess
- queryFailed
- toggleLeftBar
- queryEditorSetDb
- queryEditorSetSchema
- queryEditorSetAutorun
- queryEditorSetTitle
- queryEditorSetSql
- queryEditorSetQueryLimit
- queryEditorSetTemplateParams
```

#### Mechanism for syncing unsaved changes

![debounced](https://github.com/apache/superset/assets/1392866/a7a45d8f-14fd-40e7-a563-ce4049433c35)

There can be unsaved changes since it is syncing in a non-blocking way and asynchronously. To address this issue, unsaved changes will be stored in local storage (using localStorage in persistence mode) until they are successfully saved on the server. This hybrid approach ensures that user data is not lost and that any unsaved changes are retrieved, even in the event of an unexpected interruption.

#### Resolve conflicts

When a user opens multiple windows, they may be overwritten by the inactive editor tab state. To prevent this unexpected overwrite, the latest update time will be compared with the local cache state, and the out-of-updated state will be discarded. This ensures that users can work on multiple windows without losing their progress or having their work overwritten.

#### Following up

Creating new tabs and removing tabs are still blocking way in this commit but will be migrated into the non-blocking way in future PRs. Additionally, the existing migrateQueryEditorFromLocalStorage logic will be consolidated or replaced by the AutoSync with mutation APIs in the upcoming PRs. These changes will further improve the user experience by making the entire process non-blocking and asynchronous.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Before:

https://github.com/apache/superset/assets/1392866/7efa5a29-b7e0-4698-bb39-f4295693588d

After:

https://github.com/apache/superset/assets/1392866/3d5b2817-add5-41c8-9071-890e82e81615


### TESTING INSTRUCTIONS
Set SQLLAB_BACKEND_PERSISTENCE true and play with the SQL Lab editor and keep hard refresh to verify the persisted states.

### ADDITIONAL INFORMATION
- [x] Has associated issue: [SIP-93](https://github.com/apache/superset/issues/21385)
- [x] Required feature flags: `SQLLAB_BACKEND_PERSISTENCE`
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
feat(sqllab): non-blocking persistence mode

Commit:
remove unused ref

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/index.tsx:
```diff
@@ -0,0 +1,120 @@
+      dispatch(
+        addDangerToast(
+          t(
+            'An error occurred while storing your query in the backend. ' +
```
```suggestion
            'An error occurred while saving your editor state. ' +
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,178 @@
+  jest.useRealTimers();
+});
+
+test('sync the unsaved editor tab state with api when unsaved has made since lastest updated', async () => {
```
```suggestion
test('sync the unsaved editor tab state when there are new changes since the last update', async () => {
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,178 @@
+  fetchMock.restore();
+});
+
+test('skip syncing the unsaved editor tab state when the updates already synced', async () => {
```
```suggestion
test('skip syncing the unsaved editor tab state when the updates are already synced', async () => {
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,178 @@
+  fetchMock.restore();
+});
+
+test('renders an error toast when api update failed', async () => {
```
```suggestion
test('renders an error toast when the sync fails', async () => {
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,178 @@
+  fetchMock.restore();
+});
+
+test('skip syncing the unsaved editor tab state with api when SQLLAB_BACKEND_PERSISTENCE is off', async () => {
```
```suggestion
test('skip syncing the unsaved editor tab state when SQLLAB_BACKEND_PERSISTENCE is off', async () => {
```

michael-s-molina reviewed (commented):
Thanks for the PR @justinpark. I left some first-pass comments.

michael-s-molina commented:
@sadpandajoe @jinghua-qa could you help test the feature?

sadpandajoe commented:
> @sadpandajoe @jinghua-qa could you help test the feature?

@michael-s-molina sure, if you want to spin up the ephemeral for me. I've noticed some issues already where results aren't always showing up so i want to make sure it's not an issue between OSS and Preset

michael-s-molina commented:
/testenv up

michael-s-molina commented:
/testenv up FEATURE_SQLLAB_BACKEND_PERSISTENCE=true

sadpandajoe commented:
@justinpark looks like column auto-complete isn't working in this branch

https://github.com/apache/superset/assets/47334392/661d92f0-eede-4bb1-a10b-0749a31f8328

sadpandajoe commented:
Table schemas get removed when switching between tabs


https://github.com/apache/superset/assets/47334392/29150c3f-564b-4ae2-8257-2a6354832cdb

sadpandajoe commented:
User is not able to remove schemas without an error warning
* Select a schema in the left dropdown
* Click the remove button

What happens?
There is an error message to contact the administrator

sadpandajoe commented:
@justinpark is anything cached? I noticed that for columns it didn't show up the first time and then for another schema it works

sadpandajoe commented:
Results are not always shown with sql lab query. This is pretty hard to reproduce but also seem to be hitting a performance limit
* Create 10 sql lab tabs
* Run 10 different queries
* Go to any other page
* Go back to sql lab
* Click on each tab and rerun the query

Around tab 6 i'm starting to see performance issues with that tab loading. You can also see at the end of the video that although there is data, it won't return any results.


https://github.com/apache/superset/assets/47334392/5106c06c-4bd4-4b02-9c48-c34f7529f811

justinpark committed:
clean up constant and logic location

justinpark committed:
rename datetime to timestamp

justinpark committed:
typo

justinpark committed:
accidentally load_query_editor from new qe due to missing `loaded:true` flag

justinpark commented:
@sadpandajoe Thanks for testing. The missing columns on autocomplete has been fixed from #24611 and I also rebased this branch to include the fix. 
I also fixed the bug that is related to the tab switching with a new query editor. Could you try the test cases again?

justinpark committed:
update message in the spec

justinpark committed:
fix test for loaded flag

sadpandajoe commented:
@justinpark @michael-s-molina do we need to spin up a new ephemeral for the changes?

justinpark commented:
/testenv up FEATURE_SQLLAB_BACKEND_PERSISTENCE=true

justinpark commented:
@michael-s-molina could you help to boot up testenv?

justinpark commented:
@sadpandajoe same problem. @michael-s-molina have you faced same problem?

michael-s-molina commented:
> does ephemeral logins work for either of you @michael-s-molina / @justinpark? I haven't been testing since I can't log in.

> @sadpandajoe same problem. @michael-s-molina have you faced same problem?

Ephemeral environments are not working with `TALISMAN_ENABLED = True`. @eschutho was working on a [fix](https://github.com/apache/superset/pull/24774) but it's not merged yet. Given that `TALISMAN_ENABLED` is not a feature flag, we can't disable it so we'll need to pull this PR locally to test it.

eschutho commented:
> > does ephemeral logins work for either of you @michael-s-molina / @justinpark? I haven't been testing since I can't log in.
> 
> > @sadpandajoe same problem. @michael-s-molina have you faced same problem?
> 
> Ephemeral environments are not working with `TALISMAN_ENABLED = True`. @eschutho was working on a [fix](https://github.com/apache/superset/pull/24774) but it's not merged yet. Given that `TALISMAN_ENABLED` is not a feature flag, we can't disable it so we'll need to pull this PR locally to test it.

The talisman Pr is ready for review now.

justinpark commented:
> does ephemeral logins work for either of you michael-s-molina / justinpark? I haven't been testing since I can't log in.

@sadpandajoe could you help to test in the following env?

https://github.com/apache/superset/pull/24885#issuecomment-1665565874

sadpandajoe commented:
> > does ephemeral logins work for either of you michael-s-molina / justinpark? I haven't been testing since I can't log in.
> 
> @sadpandajoe could you help to test in the following env?
> 
> [#24885 (comment)](https://github.com/apache/superset/pull/24885#issuecomment-1665565874)

@justinpark went through and looked at the PR commented above and things seem fine to me.

justinpark committed:
update isFeatureEnabled path

justinpark committed:
ts lint

justinpark committed:
fix shouldInitialize condition

justinpark commented:
@michael-s-molina @eschutho could you review this PR again?

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,181 @@
+      sqlLab: unsavedSqlLabState,
+    },
+  });
+  await act(async () => {
```
```suggestion
  await waitFor(() => jest.runAllTimers());
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,181 @@
+      },
+    },
+  });
+  await act(async () => {
+    jest.runAllTimers();
+  });
```
```suggestion
  await waitFor(() => jest.runAllTimers());
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,181 @@
+      },
+    },
+  );
+  await act(async () => {
+    jest.runAllTimers();
+  });
```
```suggestion
  await waitFor(() => jest.runAllTimers());
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,181 @@
+      sqlLab: unsavedSqlLabState,
+    },
+  });
+  await act(async () => {
+    jest.runAllTimers();
+  });
```
```suggestion
  await waitFor(() => jest.runAllTimers());
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/index.tsx:
```diff
@@ -0,0 +1,119 @@
+  }, [debouncedUnsavedQueryEditor, dispatch, queryEditors, updateSqlEditor]);
+
+  useEffect(() => {
+    if (isError) {
```
Should this be a danger toast or just a console warning given that we're still able to save the state to local storage?

justinpark replied:
agree (I also thought same as like this). Let me change this to the console warning (or logger for monitoring purpose)

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/index.tsx:
```diff
@@ -0,0 +1,119 @@
+  setEditorTabLastUpdate,
+} from 'src/SqlLab/actions/sqlLab';
+
+const INTERVAL = 5000;
```
Maybe rename it to `SYNC_INTERVAL` to make it more descriptive?

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/EditorAutoSync.test.tsx:
```diff
@@ -0,0 +1,181 @@
+import React from 'react';
+import fetchMock from 'fetch-mock';
+import * as uiCore from '@superset-ui/core';
+import { render, act } from 'spec/helpers/testing-library';
```
```suggestion
import { render, waitFor } from 'spec/helpers/testing-library';
```

michael-s-molina commented on superset-frontend/src/SqlLab/components/EditorAutoSync/index.tsx:
```diff
@@ -0,0 +1,119 @@
+  );
+
+  useEffect(() => {
+    if (!isFeatureEnabled(FeatureFlag.SQLLAB_BACKEND_PERSISTENCE)) {
+      return;
+    }
```
This check is already done by its parent as it should given that this component does nothing without the feature flag.
```suggestion
```

michael-s-molina commented on superset-frontend/src/hooks/apiResources/sqlEditorTabs.ts:
```diff
@@ -0,0 +1,69 @@
+  endpoints: builder => ({
+    updateSqlEditorTab: builder.mutation<JsonResponse, EditorMutationParams>({
+      query: ({
+        queryEditor: {
```
Given that the query editor state is saved as a JSON object, it would be a good idea to add a version field to the object. If we have schema changes in the future, we'll be able to create transformers between the different versions and preserve compatibility. This is one of the main problems we face today with other JSON objects.

michael-s-molina commented on superset-frontend/src/SqlLab/middlewares/persistSqlLabStateEnhancer.js:
```diff
@@ -32,12 +35,47 @@ const CLEAR_ENTITY_HELPERS_MAP = {
     slicer: paths => state => {
       const subset = {};
       paths.forEach(path => {
+        if (path === PERSISTENCE_LOCAL_STORAGE_PATH) {
```
```suggestion
        if (isFeatureEnabled(FeatureFlag.SQLLAB_BACKEND_PERSISTENCE)) {
```

michael-s-molina commented on superset-frontend/src/SqlLab/middlewares/persistSqlLabStateEnhancer.js:
```diff
@@ -80,6 +118,8 @@ const sqlLabPersistStateConfig = {
 
 export const persistSqlLabStateEnhancer = persistState(
-  sqlLabPersistStateConfig.paths,
+  isFeatureEnabled(FeatureFlag.SQLLAB_BACKEND_PERSISTENCE)
+    ? [PERSISTENCE_LOCAL_STORAGE_PATH]
+    : sqlLabPersistStateConfig.paths,
```
```suggestion
  sqlLabPersistStateConfig.paths,
```

michael-s-molina commented on superset-frontend/src/SqlLab/reducers/getInitialState.test.ts:
```diff
@@ -245,4 +245,109 @@ describe('getInitialState', () => {
+                updatedAt: lastUpdatedTime + 100,
+              },
+              {
+                // out of update since last updated time is before server update time
```
```suggestion
                // no update required given that last updated time comes before server update time
```

michael-s-molina commented on superset-frontend/src/SqlLab/reducers/getInitialState.test.ts:
```diff
@@ -245,4 +245,109 @@ describe('getInitialState', () => {
+                updatedAt: lastUpdatedTime - 100,
+              },
+              {
+                // out of update since no updatedAt
```
```suggestion
                // no update required given that there's no updatedAt
```

michael-s-molina reviewed (commented):
Thanks for addressing the first pass comments @justinpark. I'm adding more comments as the result of a second pass review.

justinpark committed:
remove FeatureFlag check since upper level check exists

justinpark committed:
update comment

justinpark committed:
feature flag instead of custom key value

justinpark committed:
add version number in QueryEditor

justinpark committed:
replace error toast by logger

justinpark committed:
inspect console.warn

michael-s-molina reviewed (approved):
LGTM
